cgrok/tree.c

```diff
@@ -0,0 +1,198 @@
+  buf[size+4] = 0;
+  return buf;
+}
+
+void tctreeput(TCTREE *tree, const void *kbuf, int ksiz, const void *vbuf, int vsiz) {
+  void *key = tcpack(kbuf, ksiz);
+  void *val = tcpack(vbuf, vsiz);
+  bool inserted; 
+  void ** valPtr = dict_insert(tree->dict, key, &inserted);
+  *valPtr = val;
+}
+
+bool tctreeputkeep(TCTREE *tree, const void *kbuf, int ksiz, const void *vbuf, int vsiz) {
+  void *key = tcpack(kbuf, ksiz); 
+  void *val = tcpack(vbuf, vsiz); 
+  bool inserted;
+  void **valPtr = dict_insert(tree->dict, key, &inserted);
+  if (inserted) {
+    *valPtr = val;
+  }
+  return inserted;
+}
+
+const void *tctreeget(TCTREE *tree, const void *kbuf, int ksiz, int *sp) {
+  const void *key = tcpack(kbuf, ksiz);
+  void *value = dict_search(tree->dict, key);
+  *sp = 0;
+  if (value) {
+    *sp = *(uint32_t*)value;
+    return value+4;
+  }
+  return NULL;
+}
+
+void tctreeclear(TCTREE *tree) {
+  dict_clear(tree->dict); 
+}
+
+void tctreedel(TCTREE *tree) {
+  dict_free(tree->dict);
+  if (tree->iter) {
+    dict_itor_free(tree->iter);
+  }
+  free(tree);
+}
+
+// A simple, singly linked list
+TCLIST *tclistnew(void) {
+  TCLIST *list = malloc(sizeof(TCLIST));
+  list->len = 0;
+  TCLISTELEM *newElem = malloc(sizeof(TCLISTELEM));
+  newElem->elem = 0;
+  list->head = newElem;
+  return list;
+}
+
+int tclistnum(const TCLIST *list) {
+  return list->len;
+}
+
+void *tclistpack(void *buf, uint32_t size) {
+  void *elem = malloc(size+1);
+  memcpy(elem, buf, size);
+  ((char *)elem)[size] = 0;
+  return elem;
+}
+
+void tclistpush(TCLIST *list, const void *ptr, int size) {
+  TCLISTELEM *head = list->head;
+  for (int i=0; i < (list->len); i++) {
+    head = head->next;
+  }
+  TCLISTELEM *newElem = malloc(sizeof(TCLISTELEM));
+  newElem->elem = tclistpack(ptr, size);
+  newElem->size = size;
+  head->next = newElem;
+  list->len += 1;
+}
+
+void *tclistremove(TCLIST *list, int index, int *sp) {
+  if (index >= list->len) {
+    return NULL;
+  }
+  TCLISTELEM *head = list->head;
+  for (int i=0; i < index; i++) {
+    head = head->next;
+  }
+  
+  TCLISTELEM *removed = head->next;
+  head->next = removed->next;
+  void* elem = removed->elem;
+  *sp = removed->size;
+  free(removed);
+  list->len -= 1;
+  return elem;
+}
+
+void tclistover(TCLIST *list, int index, const void *ptr, int size) {
+  if (index >= list->len) {
+    // Out of bounds, do nothing
+    return;
+  }
+  TCLISTELEM *head = list->head;
+  for (int i=0; i <= index; i++) {
+    head = head->next;
+  }
+  free(head->elem);
+  head->elem = ptr;
+  head->size = size; 
+}
+
+const void *tclistval(const TCLIST *list, int index, int *sp) {
+  if (index >= list->len) {
+    // Out of bounds, do nothing
+    return 0;
+  }
+  TCLISTELEM *head = list->head;
+  for (int i=0; i <= index; i++) {
+    head = head->next;
+  }
+  *sp = head->size;
+  return head->elem;
+}
+
+void tclistdel(TCLIST *list) {
+  TCLISTELEM *head = list->head;
+  for (int i=0; i < list->len; i++) {
```

Comment:
nit: On line 132 you surround the `list->len` with parenthesis. Not sure if that was a style thing that didn't come through here or there's some subtle bug you were trying to avoid.